lib/src/api/servicer.rs

```diff
@@ -0,0 +1,101 @@
+use crate::api::from_proto;
+use crate::object_id::ObjectId;
+use crate::repo::ReadonlyRepo;
+use crate::settings::UserSettings;
+use config::Config;
+use itertools::Itertools;
+use jj_api::objects::{Change as ChangeProto, Workspace as WorkspaceProto};
+use jj_api::rpc::{ListWorkspacesRequest, ListWorkspacesResponse};
+use jj_lib::op_store::OperationId;
+use jj_lib::operation::Operation;
+use jj_lib::repo::RepoLoader;
+use jj_lib::workspace::WorkspaceLoader;
+
+use std::sync::Arc;
+use tonic::Status;
+
+/// The servicer handles all requests going to jj-lib. Eventually, ideally, jj-cli
+/// will interact with jj-lib purely through this class.
+pub struct Servicer {
+    default_workspace_loader: Option<WorkspaceLoader>,
+    user_settings: UserSettings,
+}
+
+impl Servicer {
+    pub fn new(default_workspace_loader: Option<WorkspaceLoader>) -> Self {
+        Self {
+            default_workspace_loader,
+            user_settings: UserSettings::from_config(Config::default()),
+        }
+    }
+
+    fn workspace_loader(
+        &self,
+        opts: &Option<jj_api::objects::RepoOptions>,
+    ) -> Result<WorkspaceLoader, Status> {
+        opts.as_ref()
+            .map(|opts| from_proto::path(&opts.repo_path))
+            .flatten()
+            .map(WorkspaceLoader::init)
+            .transpose()?
+            .or(self.default_workspace_loader.clone())
+            .ok_or_else(|| {
+                Status::invalid_argument(
+                    "No default workspace loader, and no repository.repo_path provided",
+                )
+            })
+    }
+
+    fn repo(
```

Comment:
I suspect that we may be able to take advantage of caching to improve performance of functions like this. Seems like work for later though